Source/Core/Core/HW/SI/SI_DeviceGBA.cpp

```diff
@@ -233,140 +216,149 @@ void GBASockServer::ClockSync()
-                (u8)m_recv_data[3], (u8)m_recv_data[4], num_received);
-    }
-#endif
-
-    for (size_t i = 0; i < m_recv_data.size(); i++)
-      si_buffer[i ^ 3] = m_recv_data[i];
+    for (size_t i = 0; i < recv_data.size(); i++)
+      si_buffer[i ^ 3] = recv_data[i];
   }
-
-  return (int)num_received;
-}
-
-CSIDevice_GBA::CSIDevice_GBA(SIDevices device, int device_number)
-    : ISIDevice(device, device_number), GBASockServer(device_number)
-{
+  return static_cast<int>(std::min(num_received, recv_data.size()));
 }
 
-CSIDevice_GBA::~CSIDevice_GBA()
+CSIDevice_GBA::CSIDevice_GBA(SIDevices device, int device_number) : ISIDevice(device, device_number)
 {
-  GBASockServer::Disconnect();
 }
 
 int CSIDevice_GBA::RunBuffer(u8* buffer, int length)
 {
-  if (!m_waiting_for_response)
+  switch (m_next_action)
   {
-    for (size_t i = 0; i < m_send_data.size(); i++)
-      m_send_data[i] = buffer[i ^ 3];
-
-    m_num_data_received = 0;
-    ClockSync();
-    Send(buffer);
+  case NextAction::SendCommand:
+  {
+    m_sock_server.ClockSync();
+    if (m_sock_server.Connect())
+    {
+#ifdef _DEBUG
+      NOTICE_LOG(SERIALINTERFACE, "%01d cmd %02x [> %02x%02x%02x%02x]", m_device_number, buffer[3],
+                 buffer[2], buffer[1], buffer[0], buffer[7]);
+#endif
+      m_sock_server.Send(buffer);
+    }
+    else
+    {
+      constexpr u32 reply = SI_ERROR_NO_RESPONSE;
+      std::memcpy(buffer, &reply, sizeof(reply));
+      return sizeof(reply);
+    }
+    m_last_cmd = buffer[3];
     m_timestamp_sent = CoreTiming::GetTicks();
-    m_waiting_for_response = true;
+    m_next_action = NextAction::WaitTransferTime;
   }
 
-  if (m_waiting_for_response && m_num_data_received == 0)
+  // [[fallthrough]]
+  case NextAction::WaitTransferTime:
   {
-    m_num_data_received = Receive(buffer);
+    int elapsed_time = static_cast<int>(CoreTiming::GetTicks() - m_timestamp_sent);
+    // Tell SI to ask again after TransferInterval() cycles
+    if (GetTransferTime(m_last_cmd) > elapsed_time)
+      return 0;
+    m_next_action = NextAction::ReceiveResponse;
   }
 
-  if ((GetTransferTime(m_send_data[0])) > (int)(CoreTiming::GetTicks() - m_timestamp_sent))
+  // [[fallthrough]]
+  case NextAction::ReceiveResponse:
   {
-    return 0;
+    int num_data_received = m_sock_server.Receive(buffer);
+    if (m_sock_server.IsConnected())
+    {
+#ifdef _DEBUG
+      LogTypes::LOG_LEVELS log_level = (m_last_cmd == CMD_STATUS || m_last_cmd == CMD_RESET) ?
+                                           LogTypes::LERROR :
+                                           LogTypes::LWARNING;
+      GENERIC_LOG(LogTypes::SERIALINTERFACE, log_level,
+                  "%01d                              [< %02x%02x%02x%02x%02x] (%i)",
+                  m_device_number, buffer[3], buffer[2], buffer[1], buffer[0], buffer[7],
+                  num_data_received);
+#endif
+    }
+    else
+    {
+      num_data_received = RECV_MAX_SIZE;
```

Comment:
Oh I see what happened. You put the `SI_ERROR_NO_RESPONSE` in the *send* failure path but not the *receive* failure path. It should be here, too, with a size of 4 instead of 5.